lib/spack/docs/build_systems/makefilepackage.rst

```diff
@@ -0,0 +1,295 @@
+run:
+
+.. code-block:: console
+
+   $ make
+   $ make install
+
+
+^^^^^^^^^^^^^^^
+Important files
+^^^^^^^^^^^^^^^
+
+The main file that matters for a ``MakefilePackage`` is the Makefile.
+This file will be named one of the following ways:
+
+* GNUmakefile (only works with GNU Make)
+* Makefile (most common)
+* makefile
+
+``Makefile`` is the most common name.
+
+Some Makefiles also *include* other configuration files. Check for an
+``include`` directive in the Makefile.
+
+^^^^^^^^^^^^^^^^^^^^^^^^^
+Build system dependencies
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Spack assumes that the operating system will have a valid ``make`` utility
+installed already, so you don't need to add a dependency on ``make``.
+However, if the package uses a ``GNUmakefile`` or the developers recommend
+using GNU Make, you should add a dependency on ``gmake``:
+
+.. code-block:: python
+
+   depends_on('gmake', type='build')
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+Types of Makefile packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Most of the work involved in packaging software that uses Makefiles
+involves overriding or replacing hard-coded variables. Many packages
+make the mistake of hard-coding compilers, usually for GCC or Intel.
+This is fine if you happen to be using that particular compiler, but
+Spack is designed to work with *any* compiler, and you need to ensure
+that this is the case.
+
+Depending on how the Makefile is designed, there are 4 common strategies
+that can be used to set or override the appropriate variables:
+
+"""""""""""""""""""""
+Environment variables
+"""""""""""""""""""""
+
+Make has multiple types of assignment operators. Some Makefiles
+use ``=`` to assign variables. The only way to override these
+variables is to edit the Makefile or override them on the command-line.
+However, Makefiles that use ``?=`` for assignment honor environment
+variables. Since Spack already sets ``CC``, ``CXX``, ``F77``, and ``FC``,
+you won't need to worry about setting these variables. If there are
+any other variables you need to set, you can do this in the ``edit``
+method:
+
+.. code-block:: python
+
+   def edit(self, spec, prefix):
+       env['PREFIX'] = prefix
+       env['BLASLIB'] = spec['blas'].libs.ld_flags
+
+
+``cbench`` is a good example of a simple package that does this, while
+``esmf`` is a good example of a more complex package.
+
+""""""""""""""""""""""
+Command-line arguments
+""""""""""""""""""""""
+
+If the Makefile ignores environment variables, the next thing to try
+is command-line arguments. You can do this by overriding the
+``build_targets`` attribute. If you don't need access to the spec,
+you can do this like so:
+
+.. code-block:: python
+
+   build_targets = ['CC=cc']
+
+
+If you do need access to the spec, you can create a property like so:
+
+.. code-block:: python
+
+   @property
+   def build_targets(self):
+       spec = self.spec
+
+       return [
+           'CC=cc',
+           'BLASLIB={0}'.format(spec['blas'].libs.ld_flags),
```

Comment:
I recommend using "%s"%xxx instead of "{0}".format() because (a) it fits better on flake8 lines, and (b) '%s' will always be supported in Python, it's not going away.

Reply:
`%s` has many known problems. For instance:

- It only supports ints, strings, and doubles. All other types (including the `Prefix` type) must be converted to a string first.
- `%` is a binary operator, and can only take 2 arguments.

This is why [PEP 3101](https://www.python.org/dev/peps/pep-3101) was introduced. In fact, if you read the [string formatting docs](https://docs.python.org/3.6/library/string.html), `%s` isn't even mentioned. There are currently no plans to remove it, but `.format()` is far more powerful, and supports nesting of replacement variables among other things. Also see [PEP 498](https://www.python.org/dev/peps/pep-0498/) for the future of string formatting in Python.